shopinvader_sale_coupon/services/cart.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2021 Camptocamp SA
+# @author Iván Todorovich <ivan.todorovich@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.addons.component.core import Component
+
+
+class CartService(Component):
+    _inherit = "shopinvader.cart.service"
+
+    # Public Methods
+
+    def apply_coupon(self, **params):
+        cart = self._get()
+        self._apply_coupon(cart, params)
+        return self._to_json(cart)
+
+    def recompute_coupon_lines(self, **params):
+        cart = self._get()
+        self._recompute_coupon_lines(cart, params)
+        return self._to_json(cart)
+
+    # Private Methods
+
+    def _apply_coupon(self, cart, params):
+        """Apply a coupon or promotion code.
+
+        It can raise UserError if coupon is not applicable.
+        """
+        return (
+            self.env["sale.coupon.apply.code"]
+            .with_context(active_id=cart.id)
+            .create({"coupon_code": params["code"]})
+            .process_coupon()
+        )
+
+    def _recompute_coupon_lines(self, cart, params):
+        if not params.get("skip_coupon_recompute"):
```

Comment:
you should pop this out of the params before you reach this point and call this method only if you got it.